traits/observation/observe.py

```diff
@@ -8,23 +8,38 @@
 #
 # Thanks for using Enthought open source!
 
+import asyncio
+
 from traits.observation._observe import add_or_remove_notifiers
 from traits.observation.expression import compile_expr
 
+#: Set to hold references to active async traits handlers.
+_active_handler_tasks = set()
+
 
 def dispatch_same(handler, event):
     """ Dispatch an event handler on the same thread.
 
+    This dispatcher accepts both callables and async callables, the latter
+    being dispatched asynchronously via an async Task.  Asynchronous dispatch
+    is only available when an async event loop is running; it will raise if
+    it cannot create an async Task.
+
     Parameters
     ----------
-    handler : callable(event)
+    handler : callable(event) or async callable(event)
         User-defined callable to handle change events.
         ``event`` is an object representing the change.
         Its type and content depends on the change.
     event : object
         The event object to be given to handler.
     """
-    handler(event)
+    if asyncio.iscoroutinefunction(handler):
```

Comment:
`inspect.iscoroutinefunction` may be more future-proof here - there's a possibility that `asyncio.iscoroutinefunction` will be deprecated. (See https://github.com/python/cpython/issues/94912.) There's a difference for things that use the `@coroutine` decorator, but that decorator is deprecated as of Python 3.8, and removed from Python 3.11 onwards, so I don't think the difference matters much.

Reply:
I chose the asyncio version because of this difference, but didn't realise that it was deprecated.  Will change.

Reply:
```suggestion
    if inspect.iscoroutinefunction(handler):
```